DOCS-1527,DOCS-1594: Refactor Module create page

andf-viam opened the pull request:
Refactor Create a Module page based on user feedback from several sources:
- Apply feedback from three discrete sources, listed below, plus my own awareness of "what's changed since" (example: second object in object triplet has changed context)
- Expanded seealso: examples section
- IA considerations, including:
	- Valid APIs to implement moved off-page, to parent `/registry/` overview page for now.
	- Naming your model moved off-page, to parent `/registry/` overview page for now.
	- _(Work on find/search/add/configure a module is for a different PR where I will quite possibly revisit placement of the above two.)_

### Feedback addressed in this PR:
- _(no need to read this section unless interested in what specifically was implemented, or if you were a provider of feedback and want to make sure I captured your sentiments!)_
- DOCS-1527: C++ SDK team desire to focus on extend existing API, using base API as example
	- client/server language removed.
	- all examples standardized on base example, no "your module here" or other placeholders.
- DOCS-1527: Hazal's feedback:
	- Where to add your model triplet now explicit, per language 
	- Valid API moved off-page, unique cases removed from middle of procedure, and only mentioned briefly in intro.
	- Refactored headers so that steps to perform are clearer, without duplicating concepts (create > create > create)
	- Significant clarifications to creating your model triplet, thank you Hazal, great points!!
		- Matt: this changed: Viam used to desire that the second object in the triplet was a conceptual representation of the common ground between your one (or more) models (like yahboom for both yahboom gripper and yahboom arm), but eng has recently landed on repo name instead. I removed all "like concept" stuff and focused only on repo name.
	- Links to off-docs repos as code examples now much simplified, so as to not hint that repo readmes are part of the module creation process. Hazal: you right!
	- Updated comment to go resource path `go.viam.com/rdk/examples/customresources/models/mybase` to better reflect what we're doing here. Thanks!
	- Now mentioning `go.mod` and `go.sum` with some minimal context. They are not part of module (Viam-module not Go-module) development directly, but will be seen by those inspecting other, published Go Viam modules (Go Viam!!). Good idea to include, thanks!
	- Confusion around what to do with module addressed with new Deploy section: local for Hazal's use, upload for Registry use. We should _not_ detail meta.json or executable path here, that's for the linked pages which provide that info in full context to its usage.
- DOCS-1594: Nick's feedback:
	- Constituent project files laid out directly, in intro. Right on, solid feedback Nick, thank you!
	- As with Hazal's comment on this topic: valid APIs to implement moved off-page for now. Does not belong mid-procedure. Will find the right place for it when I refactor find/search/add/configure a module section next.
	- Create a custom module now the first actionable step after intro. Bolded instruction to this affect warranted, this is a big one, agreed. Thank you!
	- HOWEVER: opted not to use generic as example because:
		- We don't yet have a generic service, Zach on it.
		- Most users, per eng SDK devs I talked to, are expected to desire to extend Viam support to new hardware that relatively closely matches an existing conceptual component.
		- if we showcase generic only, users will implement dupe functionality when they go to use generic to describe their own new gripper, when they could save time and effort by re-using most existing upstream gripper methods and just adding the new one they need, or the specific driver implementation for the new hardware they're plugging in -- which likely only needs exactly the same methods as provided upstream in the built-in gripper API for example.
		- However however, we definitely need a robust generic API example workflow as well, agreed (not in replacement though). So, when we have a generic service live, I will return to this and see if I can add a full Create a Module > Using the Generic API walkthrough, so users can select the best fit for them. I will indeed use it as teaching opportunity for "a module, all modules, include these basic building blocks" because you are correct generic you'd have to code them all, and therefore learn their utility in context to their use. Spot on, for the next effort -- will do & thank you!
		- Standardized filenames throughout for all languages
		- Clarified model namespace triplet design to land on repo-name over "conceptual parent concept", c.f. Hazal's feedback regarding same.
		- Standardized entry point file and main program, unfortunately, as both: `Entry point (main program) file` because users might be looking for either: Literal `main.cpp` program file with `main()` program (it's in the name) vs. entry point (for the module, from perspective of `viam-server`), common phrasing for, say, NPM. Subsequent references within-same-section use shortened `entry point file`. Open to feedback if both is too much, ofc.
		- Better phrasing around next steps: add a module via registry | local, right on!
		- Greatly expanded python virtual env three options with considerations of why you might want to pick one over the other, thank you!
		- Better clarity around: test your module via local (by way of Next Steps), or upload to the registry for all users.
		- Logs section moved to between model definition and main program sections. Rationale:
			- Useable in either, so between them (feedback was: after both, too far from shown example code that uses)
			- Demonstrated in sample code for model definition already, so should be placed right after.
			- New tip to indicate that said code is already included in model definition code samples above.
		- Added a link to the boost trivial logger docs, thanks!

### Questions:
- Did I get  'method' v 'function', and 'class' v 'package' correct per language (I think so!)

### TODO:
- Update module README.md files that contain links to headers changed in this PR.
	- Example: https://github.com/viamrobotics/module-example-cpp

npentrel commented on docs/registry/_index.md:
```diff
@@ -93,6 +93,60 @@ To create a new module:
+
+For example:
+
+- The `rand:yahboom:arm` model and the `rand:yahboom:gripper` model uses the repository name [yahboom](https://github.com/viam-labs/yahboom).
```
```suggestion
- The `rand:yahboom:arm` model and the `rand:yahboom:gripper` model use the repository name [yahboom](https://github.com/viam-labs/yahboom).
```

npentrel commented on docs/registry/_index.md:
````diff
@@ -93,6 +93,60 @@ To create a new module:
+  }
+  ```
+
+- The `viam-labs:audioout:pygame` model uses the repository name [audioout](https://github.com/viam-labs/audioout)
````
```suggestion
- The `viam-labs:audioout:pygame` model uses the repository name [audioout](https://github.com/viam-labs/audioout).
```

npentrel commented on docs/registry/create/_index.md:
```diff
@@ -11,154 +11,110 @@ tags:
+    "registry",
   ]
-description: "Use the Viam module system to implement modular resources that can be included in any Viam-powered machine."
+description: "Create a module to provide a new modular resource to your Viam machine."
```
```suggestion
description: "Create a module to provide a new modular resource to your machine."
```

npentrel commented on docs/registry/create/_index.md:
```diff
@@ -11,154 +11,110 @@ tags:
 
-Viam provides built-in support for a variety of different {{< glossary_tooltip term_id="component" text="components" >}} and {{< glossary_tooltip term_id="service" text="services" >}}, but you can also add support for unsupported resources by creating a module.
-A _module_ provides one or more {{< glossary_tooltip term_id="modular-resource" text="modular resources" >}}, which add support for {{< glossary_tooltip term_id="resource" text="resource" >}} {{< glossary_tooltip term_id="type" text="types" >}} or {{< glossary_tooltip term_id="model" text="models" >}} that are not built into Viam.
+Viam provides built-in support for a variety of different {{< glossary_tooltip term_id="component" text="components" >}} and {{< glossary_tooltip term_id="service" text="services" >}} -- collectively called {{< glossary_tooltip term_id="resource" text="resources" >}} -- but you might want to use Viam with a new component type or service model that isn't currently supported.
```
I'm not a huge fan of the interjecting additional information with dashes thing. I think that's generally frowned upon. Maybe we can split it into 2 sentences instead?

npentrel commented on docs/registry/create/_index.md:
```diff
@@ -11,154 +11,110 @@ tags:
-Viam provides built-in support for a variety of different {{< glossary_tooltip term_id="component" text="components" >}} and {{< glossary_tooltip term_id="service" text="services" >}}, but you can also add support for unsupported resources by creating a module.
-A _module_ provides one or more {{< glossary_tooltip term_id="modular-resource" text="modular resources" >}}, which add support for {{< glossary_tooltip term_id="resource" text="resource" >}} {{< glossary_tooltip term_id="type" text="types" >}} or {{< glossary_tooltip term_id="model" text="models" >}} that are not built into Viam.
+Viam provides built-in support for a variety of different {{< glossary_tooltip term_id="component" text="components" >}} and {{< glossary_tooltip term_id="service" text="services" >}} -- collectively called {{< glossary_tooltip term_id="resource" text="resources" >}} -- but you might want to use Viam with a new component type or service model that isn't currently supported.
+To extend support to these resources, you can create a {{< glossary_tooltip term_id="module" text="module" >}}.
```
I don't know what this sentence is supposed to mean. Maybe?
```suggestion
If the existing resources do not support your hardware or service requirements, you can create a {{< glossary_tooltip term_id="module" text="module" >}}.
```

andf-viam replied:
Trying something different, LMK what you think.

npentrel commented on docs/registry/create/_index.md:
```diff
@@ -11,154 +11,110 @@ tags:
-
-Find the subtype API as defined in the relevant <file>components/\<resource-name\>/\<resource-name>\.hpp</file> or <file>services/\<resource-name\>/<resource-name>.hpp</file> implementation (header) file in the [C++ SDK](https://github.com/viamrobotics/viam-cpp-sdk/).
+{{% alert title=Note color="note" %}}
+If you want to write a module to extend support to a new type of component that is relatively unique, and doesn't readily correspond to an existing [built-in component API](/build/program/apis/#component-apis), you may want to consider using the [generic API](/components/generic/) with your module instead of extending an existing API.
```
```suggestion
If you want to write a module to extend support to a new type of component that is relatively unique, and doesn't readily correspond to an existing [built-in component API](/build/program/apis/#component-apis), consider using the [generic API](/components/generic/) with your module instead of extending an existing API.
```

npentrel commented on docs/registry/create/_index.md:
```diff
@@ -11,154 +11,110 @@ tags:
+First, inspect the built-in class provided by the resource API that you are extending.
 
-Use the naming schema: `namespace:repo-name:name`.
+For example, if you want to add support for a new [base component](/components/base/) to Viam, you would start by looking at the built-in base component class.
```
Grammar: If sentences that use conditional clauses must use past tense (simple past or past perfect) in the if-clause of the sentence

npentrel commented on docs/registry/create/_index.md:
```diff
@@ -330,16 +287,53 @@ class MyBase(Base, Reconfigurable):
+First, inspect the built-in package provided by the resource API that you are extending.
 
-Create a folder for your module and save your code as a file named <file>my_modular_resource.go</file> inside.
+For example, if you want to add support for a new [base component](/components/base/) to Viam, you would start by looking at the built-in base component package.
```
grammar (see above)

npentrel reviewed (commented):
This looks good to me but I suggest you get Nick's and Hazal's review

andf-viam commented:
Hi @mestcihazal and @HipsterBrown -- I've refactored our module creation documentation based on yours and others' feedback -- I'd love a quick read-through from one of you to confirm that this addressed your points! Happy to hear any feedback, and make changes as needed! Thanks so much!

HipsterBrown commented on docs/registry/create/_index.md:
```diff
@@ -11,154 +11,109 @@ tags:
 
-   {{< tabs >}}
-   {{% tab name="Python" %}}
+While you can certainly combine the resource model definition and the main program code into a single file if desired (for example, a single `main.py` program that includes both the model definition and the `main()` program that uses it), this guide will use separate files for each.
```
> this guide will use separate files for each

Do we have a reason for this that we can share? Is it for clarity or separation of concerns?

andf-viam replied:
_Theory:_ The latter, great insight. My thinking:
- Keeping a clear path forward from concept to concept, as much as we can here.
- While some of our most simple examples across our repos do use one file for both 1 and 2, all of the more complicated ones use two. Starting with this file structure from the beginning might set users up for success when they later transition to module design that accomplished more complicated tasks, like a component and a service working together.
- This also helps us to demonstrate including other files and referencing the methods/functions contained within them.

_Practice:_ But, mostly, at the end of the day: our existing sample code for all three languages was split in this fashion. Refactoring to a single file for each would be much work, and for minimal gain.

But I will add context to this statement here for the end user. Thanks!

HipsterBrown reviewed (approved):
One non-blocking open question, otherwise this is a great improvement over the current "Create" docs. Thanks for the awesome refactor! 👏

stuqdog commented on docs/registry/create/_index.md:
```diff
@@ -535,18 +528,59 @@ func (b *myBase) Close(ctx context.Context) error {
+<!-- prettier-ignore -->
+| Resource Model File | Description |
+| ------------------- | ----------- |
+| [components/base/base.cpp](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/sdk/components/base/base.cpp) | Defines the specific functionality of the built-in `Base` class. |
+| [components/base/base.hpp](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/sdk/components/base/base.hpp) | Defines the implementation of functionality defined in `base.cpp`, which includes the declaration of several built-in functions such as `move_straight()`. |
```
```suggestion
| [components/base/base.hpp](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/sdk/components/base/base.hpp) | Defines the API of the built-in `Base` class, which includes the declaration of several purely virtual built-in functions such as `move_straight()`. |
| [components/base/base.cpp](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/sdk/components/base/base.cpp) | Provides implementations of the non-purely virtual functionality defined in `base.hpp`. |
```

stuqdog commented on docs/registry/create/_index.md:
```diff
@@ -535,18 +528,59 @@ func (b *myBase) Close(ctx context.Context) error {
+
+{{% alert title="Tip" color="tip" %}}
+You can view the other built-in component classes in similar fashion.
+For example, the `Camera` class is defined in [camera.cpp](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/sdk/components/camera/camera.cpp) and its functions declared in [camera.hpp](https://github.com/viamrobotics/viam-cpp-sdk/blob/main/src/viam/sdk/components/camera/camera.hpp), while the `Sensor` class is defined in [sensor.cpp](https://github.com/viamrobotics/viam-cpp-sdk/blo ...
```
(minor) I think it would help to be slightly more explicit about the fact that resource base classes are all abstract, and the resource-y methods (`stop()`, `get_image()`, etc.) are not defined in, e.g., `camera.cpp`.

stuqdog reviewed (commented):
Two small notes, nothing that need be blocking on merging!
